File: packages/cli/src/device-identity.ts
/** Same machine-id / hostname / MAC store as scripts/device_identity.py. No host Python. */

import { randomBytes } from "crypto";
import fs from "fs";
import path from "path";

export type DeviceIdentity = {
  machineId: string;
  hostname: string;
  mac: string;
};

const MACHINE_ID_RE = /^[0-9a-f]{32}$/;
const HOSTNAME_RE = /^[a-z0-9]([a-z0-9-]*[a-z0-9])?$/;
const MAC_RE = /^[0-9a-f]{2}(:[0-9a-f]{2}){5}$/;
const PREFIXES = ["lenovo-pc", "honor-pc", "xiaomi-pc", "asus-pc", "dell-pc", "hp-pc", "thinkpad"];
const ENV_NAME = "device-identity.env";
const JSON_NAME = "device-identity.json";
const ENV_TEMP_RE = /^device-identity\.env\.(?:\d+\.[0-9a-f]{8}|[A-Za-z0-9_-]+)$/;

function hasLineBreak(value: string): boolean {
  return value.includes("\n") || value.includes("\r");
}

export function validMachineId(value: string): boolean {
  return !hasLineBreak(value) && MACHINE_ID_RE.test(value);
}

export function validHostname(value: string): boolean {
  return (
    !hasLineBreak(value) &&
    value.length >= 1 &&
    value.length <= 63 &&
    HOSTNAME_RE.test(value)
  );
}

export function validMac(value: string): boolean {
  if (hasLineBreak(value) || !MAC_RE.test(value)) return false;
  return parseInt(value.slice(0, 2), 16) % 2 === 0;
}

export function parseDeviceIdentity(raw: unknown): DeviceIdentity | null {
  if (!raw || typeof raw !== "object") return null;
  const rec = raw as Record<string, unknown>;
  if (typeof rec.machineId !== "string" || !validMachineId(rec.machineId)) return null;
  if (typeof rec.hostname !== "string" || !validHostname(rec.hostname)) return null;
  if (typeof rec.mac !== "string" || !validMac(rec.mac)) return null;
  return { machineId: rec.machineId, hostname: rec.hostname, mac: rec.mac };
}

export function generateDeviceIdentity(machineId: string): DeviceIdentity {
  const prefix = PREFIXES[parseInt(machineId.slice(0, 2), 16) % PREFIXES.length];
  const num = (parseInt(machineId.slice(2, 6), 16) % 900) + 100;
  const identity: DeviceIdentity = {
    machineId,
    hostname: `${prefix}-${num}`,
    mac: ["00", "1b", "21", machineId.slice(6, 8), machineId.slice(8, 10), machineId.slice(10, 12)].join(":"),
  };
  if (!parseDeviceIdentity(identity)) {
    throw new Error("generated device identity failed validation");
  }
  return identity;
}

type PathKind = "absent" | "symlink" | "file" | "dir" | "other";

function pathKind(target: string): PathKind {
  let st: fs.Stats;
  try {
    st = fs.lstatSync(target);
  } catch {
    return "absent";
  }
  if (st.isSymbolicLink()) return "symlink";
  if (st.isFile()) return "file";
  if (st.isDirectory()) return "dir";
  return "other";
}

function requireAbsentOrRegular(target: string): void {
  const kind = pathKind(target);
  if (kind === "symlink") throw new Error(`${target} is a symlink`);
  if (kind === "other" || kind === "dir") throw new Error(`${target} is not a regular file`);
}

function hardenRegular(target: string): void {
  requireAbsentOrRegular(target);
  const flags = fs.constants.O_RDONLY | (fs.constants.O_NOFOLLOW ?? 0);
  const fd = fs.openSync(target, flags);
  try {
    let st = fs.fstatSync(fd);
    if (st.nlink > 1) {
      cleanupOwnedTempLinks(target, st);
      st = fs.fstatSync(fd);
    }
    if (st.nlink !== 1) throw new Error(`${target} has unexpected link count ${st.nlink}`);
    if (typeof process.getuid === "function") {
      const uid = process.getuid();
      if (uid !== 0 && st.uid !== uid) {
        throw new Error(`${target} is not owned by the current user`);
      }
    }
    fs.fchmodSync(fd, 0o600);
  } finally {
    fs.closeSync(fd);
  }
}

function cleanupOwnedTempLinks(target: string, targetStat: fs.Stats): void {
  const dir = path.dirname(target);
  const base = path.basename(target);
  for (const name of fs.readdirSync(dir)) {
    if (name === base || !ENV_TEMP_RE.test(name)) continue;
    const candidate = path.join(dir, name);
    let st: fs.Stats;
    try {
      st = fs.lstatSync(candidate);
    } catch {
      continue;
    }
    if (
      st.isFile() &&
      st.dev === targetStat.dev &&
      st.ino === targetStat.ino &&
      st.nlink === targetStat.nlink
    ) {
      fs.unlinkSync(candidate);
    }
  }
}

function parseEnvFile(target: string): DeviceIdentity {
  requireAbsentOrRegular(target);
  const data = fs.readFileSync(target);
  if (data.includes(0)) throw new Error(`NUL in ${target}`);
  if (data.some((byte) => byte > 0x7f)) throw new Error(`non-ascii identity in ${target}`);
  const text = data.toString("utf-8");
  let machineId: string | undefined;
  let hostname: string | undefined;
  let mac: string | undefined;
  for (const line of text.split(/\n/)) {
    if (line === "") continue;
    if (line.startsWith("AGENT_WECHAT_MACHINE_ID=")) {
      if (machineId) throw new Error(`duplicate AGENT_WECHAT_MACHINE_ID in ${target}`);
      machineId = line.slice("AGENT_WECHAT_MACHINE_ID=".length);
    } else if (line.startsWith("AGENT_WECHAT_HOSTNAME=")) {
      if (hostname) throw new Error(`duplicate AGENT_WECHAT_HOSTNAME in ${target}`);
      hostname = line.slice("AGENT_WECHAT_HOSTNAME=".length);
    } else if (line.startsWith("AGENT_WECHAT_MAC=")) {
      if (mac) throw new Error(`duplicate AGENT_WECHAT_MAC in ${target}`);
      mac = line.slice("AGENT_WECHAT_MAC=".length);
    } else {
      throw new Error(`unexpected line in ${target}`);
    }
  }
  const parsed = parseDeviceIdentity({ machineId, hostname, mac });
  if (!parsed) throw new Error(`incomplete identity in ${target}`);
  return parsed;
}

function parseJsonFile(target: string): DeviceIdentity {
  requireAbsentOrRegular(target);
  const data = fs.readFileSync(target);
  if (data.includes(0)) throw new Error(`NUL in ${target}`);
  const parsed = parseDeviceIdentity(JSON.parse(data.toString("utf-8")));
  if (!parsed) throw new Error(`invalid JSON identity in ${target}`);
  return parsed;
}

function exclusivePublish(envFile: string, identity: DeviceIdentity): boolean {
  requireAbsentOrRegular(envFile);
  const tmp = envFile + `.${process.pid}.${randomBytes(4).toString("hex")}`;
  const payload = [
    `AGENT_WECHAT_MACHINE_ID=${identity.machineId}`,
    `AGENT_WECHAT_HOSTNAME=${identity.hostname}`,
    `AGENT_WECHAT_MAC=${identity.mac}`,
    "",
  ].join("\n");
  fs.writeFileSync(tmp, payload, { mode: 0o600, flag: "wx" });
  try {
    fs.linkSync(tmp, envFile);
  } catch {
    fs.unlinkSync(tmp);
    return false;
  }
  fs.unlinkSync(tmp);
  fs.chmodSync(envFile, 0o600);
  return true;
}

function prepareIdentDir(dir: string): void {
  const kind = pathKind(dir);
  if (kind === "symlink") throw new Error(`${dir} is a symlink`);
  if (kind === "file" || kind === "other") throw new Error(`${dir} is not a directory`);
  if (kind === "absent") fs.mkdirSync(dir, { recursive: true, mode: 0o700 });
  try {
    fs.chmodSync(dir, 0o700);
  } catch {
    // best effort
  }
}

function sameIdentity(a: DeviceIdentity, b: DeviceIdentity): boolean {
  return a.machineId === b.machineId && a.hostname === b.hostname && a.mac === b.mac;
}

/** Serialized first-run via exclusive publish of device-identity.env. */
export function ensureDeviceIdentity(dir: string): DeviceIdentity {
  prepareIdentDir(dir);
  const envFile = path.join(dir, ENV_NAME);
  const jsonFile = path.join(dir, JSON_NAME);
  requireAbsentOrRegular(envFile);
  requireAbsentOrRegular(jsonFile);

  const envId = pathKind(envFile) === "file" ? parseEnvFile(envFile) : null;
  const jsonId = pathKind(jsonFile) === "file" ? parseJsonFile(jsonFile) : null;
  if (envId && jsonId && !sameIdentity(envId, jsonId)) {
    throw new Error(`conflicting ${ENV_NAME} and ${JSON_NAME} in ${dir}`);
  }
  if (envId) {
    hardenRegular(envFile);
    return envId;
  }
  if (jsonId) {
    if (exclusivePublish(envFile, jsonId)) return parseEnvFile(envFile);
    const winner = parseEnvFile(envFile);
    if (!sameIdentity(winner, jsonId)) {
      throw new Error(`conflicting ${ENV_NAME} and ${JSON_NAME} in ${dir}`);
    }
    return winner;
  }

  const generated = generateDeviceIdentity(randomBytes(16).toString("hex"));
  if (exclusivePublish(envFile, generated)) return parseEnvFile(envFile);
  return parseEnvFile(envFile);
}

export function buildDockerRunArgs(
  identity: DeviceIdentity,
  opts: {
    image: string;
    containerName: string;
    tokenPath: string;
    port: number;
    proxy?: string;
  },
): string[] {
  const args = [
    "run",
    "-d",
    "--name",
    opts.containerName,
    "--hostname",
    identity.hostname,
    "--mac-address",
    identity.mac,
    "--security-opt",
    "seccomp=unconfined",
    "--cap-add=SYS_PTRACE",
    "--cap-add=NET_ADMIN",
    "-p",
    `${opts.port}:${opts.port}`,
    "-v",
    `${opts.containerName}-data:/data`,
    "-v",
    `${opts.containerName}-wechat-home:/home/wechat`,
    "-v",
    `${opts.tokenPath}:/data/auth-token:ro`,
    "-e",
    `AGENT_WECHAT_MACHINE_ID=${identity.machineId}`,
    "-e",
    `AGENT_WECHAT_HOSTNAME=${identity.hostname}`,
    "-e",
    `AGENT_WECHAT_MAC=${identity.mac}`,
  ];
  if (opts.proxy) {
    args.push("-e", `PROXY=${opts.proxy}`);
  }
  args.push(opts.image);
  return args;
}

File: scripts/device_identity.py
#!/usr/bin/env python3
"""Canonical host-side device identity. CLI and Compose both use this file."""

from __future__ import annotations

import fcntl
import json
import os
import re
import shlex
import stat
import sys
import tempfile
import time
import uuid
from typing import Callable, Optional

PREFIXES = ("lenovo-pc", "honor-pc", "xiaomi-pc", "asus-pc", "dell-pc", "hp-pc", "thinkpad")
MACHINE_ID_RE = re.compile(r"^[0-9a-f]{32}$")
HOSTNAME_RE = re.compile(r"^[a-z0-9]([a-z0-9-]*[a-z0-9])?$")
MAC_RE = re.compile(r"^[0-9a-f]{2}(:[0-9a-f]{2}){5}$")
ENV_NAME = "device-identity.env"
JSON_NAME = "device-identity.json"
LOCK_NAME = ".device-identity.lock"
ENV_TEMP_RE = re.compile(r"^device-identity\.env\.(?:\d+\.[0-9a-f]{8}|[A-Za-z0-9_-]+)$")


def die(message: str) -> None:
    print(f"[identity] {message}", file=sys.stderr)
    raise SystemExit(1)


def valid_machine_id(value: str) -> bool:
    return "\n" not in value and "\r" not in value and bool(MACHINE_ID_RE.fullmatch(value))


def valid_hostname(value: str) -> bool:
    return (
        "\n" not in value
        and "\r" not in value
        and 1 <= len(value) <= 63
        and bool(HOSTNAME_RE.fullmatch(value))
    )


def valid_mac(value: str) -> bool:
    if "\n" in value or "\r" in value or not MAC_RE.fullmatch(value):
        return False
    return int(value.split(":", 1)[0], 16) % 2 == 0


def validate_override(name: str, value: Optional[str], checker: Callable[[str], bool]) -> Optional[str]:
    if not value:
        return None
    if not checker(value):
        die(f"invalid {name} override")
    return value


def path_kind(path: str) -> str:
    try:
        info = os.lstat(path)
    except FileNotFoundError:
        return "absent"
    if stat.S_ISLNK(info.st_mode):
        return "symlink"
    if stat.S_ISREG(info.st_mode):
        return "file"
    if stat.S_ISDIR(info.st_mode):
        return "dir"
    return "other"


def require_absent_or_regular(path: str) -> None:
    kind = path_kind(path)
    if kind == "symlink":
        die(f"{path} is a symlink")
    if kind == "other":
        die(f"{path} is not a regular file")


def parse_env_identity(path: str) -> tuple[str, str, str]:
    require_absent_or_regular(path)
    data = open(path, "rb").read()
    if b"\0" in data:
        die(f"NUL in {path}")
    try:
        text = data.decode("ascii")
    except UnicodeDecodeError:
        die(f"non-ascii identity in {path}")
    mid = hn = mac = None
    for line in text.splitlines():
        if line.startswith("AGENT_WECHAT_MACHINE_ID="):
            if mid is not None:
                die(f"duplicate AGENT_WECHAT_MACHINE_ID in {path}")
            value = line[len("AGENT_WECHAT_MACHINE_ID=") :]
            if not valid_machine_id(value):
                die(f"invalid AGENT_WECHAT_MACHINE_ID in {path}")
            mid = value
        elif line.startswith("AGENT_WECHAT_HOSTNAME="):
            if hn is not None:
                die(f"duplicate AGENT_WECHAT_HOSTNAME in {path}")
            value = line[len("AGENT_WECHAT_HOSTNAME=") :]
            if not valid_hostname(value):
                die(f"invalid AGENT_WECHAT_HOSTNAME in {path}")
            hn = value
        elif line.startswith("AGENT_WECHAT_MAC="):
            if mac is not None:
                die(f"duplicate AGENT_WECHAT_MAC in {path}")
            value = line[len("AGENT_WECHAT_MAC=") :]
            if not valid_mac(value):
                die(f"invalid AGENT_WECHAT_MAC in {path}")
            mac = value
        else:
            die(f"unexpected line in {path}")
    if not (mid and hn and mac):
        die(f"incomplete identity in {path}")
    return mid, hn, mac


def parse_json_identity(path: str) -> tuple[str, str, str]:
    require_absent_or_regular(path)
    raw = open(path, "rb").read()
    if b"\0" in raw:
        die(f"NUL in {path}")
    try:
        data = json.loads(raw.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError):
        die(f"invalid JSON identity in {path}")
    if not isinstance(data, dict):
        die(f"invalid JSON identity in {path}")
    mid = data.get("machineId")
    hn = data.get("hostname")
    mac = data.get("mac")
    if not isinstance(mid, str) or not valid_machine_id(mid):
        die(f"invalid machineId in {path}")
    if not isinstance(hn, str) or not valid_hostname(hn):
        die(f"invalid hostname in {path}")
    if not isinstance(mac, str) or not valid_mac(mac):
        die(f"invalid mac in {path}")
    return mid, hn, mac


def derive_hostname(mid: str) -> str:
    idx = int(mid[0:2], 16) % len(PREFIXES)
    num = int(mid[2:6], 16) % 900 + 100
    return f"{PREFIXES[idx]}-{num}"


def derive_mac(mid: str) -> str:
    return f"00:1b:21:{mid[6:8]}:{mid[8:10]}:{mid[10:12]}"


def exclusive_publish(env_file: str, mid: str, hn: str, mac: str) -> bool:
    directory = os.path.dirname(env_file)
    require_absent_or_regular(env_file)
    fd, tmp = tempfile.mkstemp(prefix="device-identity.env.", dir=directory)
    try:
        os.fchmod(fd, 0o600)
        payload = (
            f"AGENT_WECHAT_MACHINE_ID={mid}\n"
            f"AGENT_WECHAT_HOSTNAME={hn}\n"
            f"AGENT_WECHAT_MAC={mac}\n"
        ).encode("ascii")
        os.write(fd, payload)
        os.fsync(fd)
    finally:
        os.close(fd)
    try:
        os.link(tmp, env_file)
    except FileExistsError:
        os.unlink(tmp)
        return False
    os.unlink(tmp)
    os.chmod(env_file, 0o600)
    return True


def harden_regular(path: str) -> None:
    require_absent_or_regular(path)
    flags = os.O_RDONLY
    if hasattr(os, "O_NOFOLLOW"):
        flags |= os.O_NOFOLLOW
    fd = os.open(path, flags)
    try:
        info = os.fstat(fd)
        if info.st_nlink > 1:
            cleanup_owned_temp_links(path, info)
            info = os.fstat(fd)
        if info.st_nlink != 1:
            die(f"{path} has unexpected link count {info.st_nlink}")
        if info.st_uid != os.getuid() and os.getuid() != 0:
            die(f"{path} is not owned by the current user")
        os.fchmod(fd, 0o600)
    finally:
        os.close(fd)


def cleanup_owned_temp_links(path: str, target_info: os.stat_result) -> None:
    directory = os.path.dirname(path)
    base = os.path.basename(path)
    for name in os.listdir(directory):
        if name == base or not ENV_TEMP_RE.fullmatch(name):
            continue
        candidate = os.path.join(directory, name)
        try:
            info = os.lstat(candidate)
        except FileNotFoundError:
            continue
        if (
            stat.S_ISREG(info.st_mode)
            and info.st_dev == target_info.st_dev
            and info.st_ino == target_info.st_ino
            and info.st_nlink == target_info.st_nlink
        ):
            os.unlink(candidate)


def acquire_lock(lock_path: str) -> int:
    require_absent_or_regular(lock_path)
    flags = os.O_RDWR | os.O_CREAT
    if hasattr(os, "O_NOFOLLOW"):
        flags |= os.O_NOFOLLOW
    try:
        fd = os.open(lock_path, flags, 0o600)
    except OSError as exc:
        die(f"cannot open lock {lock_path}: {exc}")
    try:
        info = os.fstat(fd)
        if not stat.S_ISREG(info.st_mode):
            die(f"{lock_path} is not a regular file")
        fcntl.flock(fd, fcntl.LOCK_EX)
    except Exception:
        os.close(fd)
        raise
    return fd


def prepare_ident_dir(ident_dir: str) -> None:
    kind = path_kind(ident_dir)
    if kind == "symlink":
        die(f"{ident_dir} is a symlink")
    if kind == "file":
        die(f"{ident_dir} is not a directory")
    if kind == "absent":
        os.makedirs(ident_dir, mode=0o700, exist_ok=True)
    elif kind != "dir":
        die(f"{ident_dir} is not a directory")
    try:
        os.chmod(ident_dir, 0o700)
    except OSError:
        pass


def load_or_create(ident_dir: str) -> tuple[str, str, str]:
    prepare_ident_dir(ident_dir)
    env_file = os.path.join(ident_dir, ENV_NAME)
    json_file = os.path.join(ident_dir, JSON_NAME)
    require_absent_or_regular(env_file)
    require_absent_or_regular(json_file)

    env_id = parse_env_identity(env_file) if path_kind(env_file) == "file" else None
    json_id = parse_json_identity(json_file) if path_kind(json_file) == "file" else None
    if env_id and json_id and env_id != json_id:
        die(f"conflicting {ENV_NAME} and {JSON_NAME} in {ident_dir}")
    if env_id:
        harden_regular(env_file)
        return env_id
    if json_id:
        if exclusive_publish(env_file, *json_id):
            return parse_env_identity(env_file)
        env_id = parse_env_identity(env_file)
        if env_id != json_id:
            die(f"conflicting {ENV_NAME} and {JSON_NAME} in {ident_dir}")
        return env_id

    mid = hn = mac = None
    if os.environ.get("AGENT_WECHAT_IDENTITY_FROM_ENV") == "1":
        mid = validate_override(
            "AGENT_WECHAT_MACHINE_ID",
            os.environ.get("AGENT_WECHAT_MACHINE_ID"),
            valid_machine_id,
        )
        hn = validate_override(
            "AGENT_WECHAT_HOSTNAME",
            os.environ.get("AGENT_WECHAT_HOSTNAME"),
            valid_hostname,
        )
        mac = validate_override(
            "AGENT_WECHAT_MAC",
            os.environ.get("AGENT_WECHAT_MAC"),
            valid_mac,
        )
    if mid is None:
        mid = uuid.uuid4().hex
    if not valid_machine_id(mid):
        die("generated machine-id is invalid")
    if hn is None:
        hn = derive_hostname(mid)
    if mac is None:
        mac = derive_mac(mid)
    if not valid_hostname(hn):
        die("generated hostname is invalid")
    if not valid_mac(mac):
        die("generated MAC is invalid")
    if exclusive_publish(env_file, mid, hn, mac):
        return parse_env_identity(env_file)
    return parse_env_identity(env_file)


def main() -> None:
    ident_dir = (
        sys.argv[1]
        if len(sys.argv) > 1
        else os.environ.get("AGENT_WECHAT_IDENTITY_DIR")
        or os.path.join(os.path.expanduser("~"), ".config", "agent-wechat")
    )
    prepare_ident_dir(ident_dir)
    lock_path = os.path.join(ident_dir, LOCK_NAME)
    lock_fd = acquire_lock(lock_path)
    try:
        hold_ms = int(os.environ.get("AGENT_WECHAT_IDENTITY_HOLD_MS") or "0")
    except ValueError:
        hold_ms = 0
    try:
        mid, hn, mac = load_or_create(ident_dir)
        if hold_ms > 0:
            time.sleep(hold_ms / 1000.0)
    finally:
        os.close(lock_fd)
    print(f"export AGENT_WECHAT_MACHINE_ID={shlex.quote(mid)}")
    print(f"export AGENT_WECHAT_HOSTNAME={shlex.quote(hn)}")
    print(f"export AGENT_WECHAT_MAC={shlex.quote(mac)}")


if __name__ == "__main__":
    main()
